File: tests/test_proxy_real.py
"""Real proxy rotation tests — launches browsers with actual proxies.

Run: pytest tests/test_proxy_real.py -v -m slow
"""
import pytest
from cloakbrowser import ProxyRotator, launch

PROXIES = [
    "http://user:pass@proxy1.example.com:5610",
    "http://user:pass@proxy2.example.com:4586",
    "http://user:pass@proxy3.example.com:5906",
]


@pytest.mark.slow
class TestRoundRobin:
    def test_each_proxy_returns_unique_ip(self):
        rotator = ProxyRotator(proxies=PROXIES, strategy="round_robin")
        ips = []
        for i in range(3):
            proxy = rotator.next()
            browser = launch(headless=True, proxy=proxy)
            page = browser.new_page()
            page.goto("https://api.ipify.org?format=json", timeout=15000)
            ip = page.text_content("body")
            ips.append(ip)
            rotator.report_success(proxy)
            browser.close()
        assert len(set(ips)) == 3, f"Expected 3 unique IPs, got {set(ips)}"


@pytest.mark.slow
class TestSticky:
    def test_sticky_reuses_same_proxy(self):
        sticky = ProxyRotator(
            proxies=PROXIES[:2],
            strategy="round_robin",
            sticky_count=2,
        )
        ips = []
        for _ in range(4):
            proxy = sticky.next()
            browser = launch(headless=True, proxy=proxy)
            page = browser.new_page()
            page.goto("https://api.ipify.org?format=json", timeout=15000)
            ip = page.text_content("body")
            ips.append(ip)
            sticky.report_success(proxy)
            browser.close()
        assert ips[0] == ips[1], "Sticky: first 2 should match"
        assert ips[2] == ips[3], "Sticky: last 2 should match"
        assert ips[0] != ips[2], "Sticky: pairs should differ"


@pytest.mark.slow
class TestSession:
    def test_session_tracks_success(self):
        tracker = ProxyRotator(proxies=PROXIES, strategy="least_failures")
        with tracker.session() as proxy:
            browser = launch(headless=True, proxy=proxy)
            page = browser.new_page()
            page.goto("https://api.ipify.org?format=json", timeout=15000)
            ip = page.text_content("body")
            assert ip, "Should return IP"
            browser.close()
        stats = tracker.stats()
        total_fails = sum(s["fail_count"] for s in stats)
        assert total_fails == 0, "No failures expected"


class TestSocks5Validation:
    def test_socks5_with_auth_raises(self):
        with pytest.raises(ValueError, match="SOCKS5 with authentication"):
            ProxyRotator(["socks5://user:pass@host:1080"])

    def test_socks5_without_auth_accepted(self):
        r = ProxyRotator(["socks5://host:1080"])
        assert len(r) == 1

    def test_socks5_dict_with_auth_raises(self):
        with pytest.raises(ValueError, match="SOCKS5 with authentication"):
            ProxyRotator([{"server": "socks5://host:1080", "username": "u", "password": "p"}])

    def test_add_socks5_with_auth_raises(self):
        r = ProxyRotator(["http://proxy:8080"])
        with pytest.raises(ValueError, match="SOCKS5 with authentication"):
            r.add("socks5://user:pass@host:1080")


if __name__ == "__main__":
    import sys
    sys.exit(pytest.main([__file__, "-v", "--tb=short", "-x"]))

File: tests/test_proxy_real.mjs
// tests/test_proxy_real.mjs
/**
 * Real proxy rotation test (JS) — launches browsers with actual proxies.
 * Run: node tests/test_proxy_real.mjs
 */
import { launch } from '../js/dist/index.js';
import { ProxyRotator } from '../js/dist/proxy-rotator.js';

const PROXIES = [
  'http://user:pass@proxy1.example.com:5610',
  'http://user:pass@proxy2.example.com:4586',
  'http://user:pass@proxy3.example.com:5906',
];

const results = [];

function check(name, passed, detail = '') {
  const status = passed ? 'PASS' : 'FAIL';
  let msg = `  [${status}] ${name}`;
  if (detail) msg += ` — ${detail}`;
  console.log(msg);
  results.push({ name, status });
}

async function main() {
  console.log('='.repeat(70));
  console.log('  PROXY ROTATION — REAL BROWSER TEST (JS)');
  console.log('='.repeat(70));

  // ---- Test 1: Round-robin — each proxy returns different IP ----
  console.log('\n--- Test 1: Round-robin rotation ---');
  const rotator = new ProxyRotator(PROXIES, { strategy: 'round_robin' });
  const ips = [];

  for (let i = 0; i < 3; i++) {
    const proxy = rotator.next();
    const browser = await launch({ headless: true, proxy });
    const page = await browser.newPage();
    await page.goto('https://api.ipify.org?format=json', { timeout: 15000 });
    const ip = await page.textContent('body');
    console.log(`  Proxy ${i + 1}: ${ip}`);
    ips.push(ip);
    rotator.reportSuccess(proxy);
    await browser.close();
  }

  const unique = new Set(ips);
  check('round-robin unique IPs', unique.size === 3, `${unique.size} unique`);

  // ---- Test 2: Sticky — same IP for 2 requests ----
  console.log('\n--- Test 2: Sticky session (2 requests) ---');
  const sticky = new ProxyRotator(PROXIES.slice(0, 2), {
    strategy: 'round_robin',
    stickyCount: 2,
  });
  const stickyIps = [];

  for (let i = 0; i < 4; i++) {
    const proxy = sticky.next();
    const browser = await launch({ headless: true, proxy });
    const page = await browser.newPage();
    await page.goto('https://api.ipify.org?format=json', { timeout: 15000 });
    const ip = await page.textContent('body');
    console.log(`  Request ${i + 1}: ${ip}`);
    stickyIps.push(ip);
    sticky.reportSuccess(proxy);
    await browser.close();
  }

  check('sticky first pair', stickyIps[0] === stickyIps[1], `${stickyIps[0]} == ${stickyIps[1]}`);
  check('sticky second pair', stickyIps[2] === stickyIps[3], `${stickyIps[2]} == ${stickyIps[3]}`);
  check('sticky pairs differ', stickyIps[0] !== stickyIps[2], `${stickyIps[0]} != ${stickyIps[2]}`);

  // ---- Test 3: withSession — success tracking ----
  console.log('\n--- Test 3: withSession tracking ---');
  const tracker = new ProxyRotator(PROXIES, { strategy: 'least_failures' });

  await tracker.withSession(async (proxy) => {
    const browser = await launch({ headless: true, proxy });
    const page = await browser.newPage();
    await page.goto('https://api.ipify.org?format=json', { timeout: 15000 });
    console.log(`  Session proxy: ${await page.textContent('body')}`);
    await browser.close();
  });

  const stats = tracker.stats();
  const totalFails = stats.reduce((sum, s) => sum + s.failCount, 0);
  check('session no failures', totalFails === 0, `total fails: ${totalFails}`);
  for (const s of stats) {
    console.log(`  ${s.proxy}: uses=${s.useCount}, fails=${s.failCount}`);
  }

  // ---- Test 4: SOCKS5 with auth rejected ----
  console.log('\n--- Test 4: SOCKS5 auth validation ---');
  let socks5Rejected = false;
  try {
    new ProxyRotator(['socks5://user:pass@host:1080']);
  } catch (e) {
    socks5Rejected = e.message.includes('SOCKS5');
  }
  check('socks5 with auth rejected', socks5Rejected);

  let socks5NoAuth = false;
  try {
    new ProxyRotator(['socks5://host:1080']);
    socks5NoAuth = true;
  } catch (_) {}
  check('socks5 without auth accepted', socks5NoAuth);

  let socks5DictRejected = false;
  try {
    new ProxyRotator([{ server: 'socks5://host:1080', username: 'u', password: 'p' }]);
  } catch (e) {
    socks5DictRejected = e.message.includes('SOCKS5');
  }
  check('socks5 dict with auth rejected', socks5DictRejected);

  let addRejected = false;
  try {
    const r = new ProxyRotator(['http://proxy:8080']);
    r.add('socks5://user:pass@host:1080');
  } catch (e) {
    addRejected = e.message.includes('SOCKS5');
  }
  check('add socks5 with auth rejected', addRejected);

  // ---- Summary ----
  console.log('\n' + '='.repeat(70));
  console.log('  SUMMARY');
  console.log('='.repeat(70));

  const passed = results.filter(r => r.status === 'PASS').length;
  const failed = results.filter(r => r.status === 'FAIL').length;

  for (const r of results) {
    const icon = r.status === 'PASS' ? 'OK' : 'XX';
    console.log(`  [${icon}] ${r.name}`);
  }

  console.log(`\n  ${passed}/${results.length} passed, ${failed} failed`);
  if (failed === 0) console.log('  *** ALL TESTS PASSED ***');
  console.log('='.repeat(70));

  await new Promise(r => setTimeout(r, 500));
  process.exit(failed > 0 ? 1 : 0);
}

main().catch(e => { console.error(e); process.exit(1); });
